### ISSUE TITLE
How to control the fields defined in proto to remain the same as the output fields in typescript file

### ISSUE THREAD

**Issue by imxood** [timostamm/protobuf-ts#177]
How to control the fields defined in proto to remain the same as the output fields in typescript file

![image](https://user-images.githubusercontent.com/14890680/138586075-c007e07e-8222-492f-b844-b204340ada73.png)

**Comment by kskalski** [comment]
Just to throw in my 2 cents - there are many options that influence how the code is generated or transformed to JSON (including casing of fields for the latter). I think it should be up to the developers to decide which naming conventions they want to use and there are some important arguments on why somebody would want to keep the original names:
- least surprise principle, you define field ``FooBar`` or ``waz_baz`` and you get the same thing you defined in the generated code
- ability to quickly search for the field in the codebase, you see field ``some_number`` and you can just grep the code for it instead of manually transforming it to whatever casing given output language uses (it's not a problem if you only use one language, since then you can adopt naming convention of that language to proto)
- using different casing in each output language makes it harder to do code sharing by means of transforming / generating / manually copying pieces of common code across the languages/projects (this is my, maybe obsure, but real use-case: I do have a library of common algorithms operating on my models written in C#, which I transform to TypeScript by third party tool and of course dealing with varying naming conventions would be a pain)
- although it seems CamelCase is indeed widely adopted convention for TypeScript, I'm not sure every developer uses it (I don't) or should be forced to use it. Some people might prefer uniformity (e.g. using single naming convention across whole project independent of the languages) over adherence to whatever background given language's community had when creating that language

The fact that original protoc generators do not have such option is probably the biggest argument against it, so I'm surely going to accept whatever opinionated decision you as the author of this plugin take. However, given richness of this library and myriad of options (e.g. ``keep_enum_prefix``) already added, the ability to preserve original names seems like a big omission.

**Comment by timostamm** [comment]
Hey Kamil, 

most parts of protobuf-ts operate on generated metadata. This is how we can achieve [10% of the code size of google-protobuf](https://github.com/timostamm/protobuf-ts/blob/master/MANUAL.md#code-size-vs-speed).

"name" in this example is the original field name (needed for the JSON format):

```ts
        super("spec.ScalarValuesMessage", [
            { no: 1, name: "double_field", kind: "scalar", T: 1 /*ScalarType.DOUBLE*/ },
            { no: 2, name: "float_field", kind: "scalar", T: 2 /*ScalarType.FLOAT*/ },
        // ...
```

At runtime, we need to access properties of the message interface `Person` based on the field metadata:

```ts
export interface ScalarValuesMessage {
    /**
     * @generated from protobuf field: double double_field = 1;
     */
    doubleField: number;
    /**
     * @generated from protobuf field: float float_field = 2;
     */
    floatField: number;
    // ...
```

Field metadata actually has a second property "localName". The contract is that "localName" must match the interface property names. We do _not_ need to specify "localName" for ScalarValuesMessage, because "localName" is trivially inferred from "name" through `lowerCamelCase()`.

There are a few exceptions where "localName" is required, because there are some property names such as `toJSON` that would not be safe to use. These names are escaped by the code generator, and added as "localName" so that the runtime does not need the escaping logic.

A plugin option to use original field names for properties could use "localName", but most generated code would look like this:

```ts
        super("spec.ScalarValuesMessage", [
            { no: 1, name: "double_field", localName: "double_field", kind: "scalar", T: 1 /*ScalarType.DOUBLE*/ },
            { no: 2, name: "float_field", localName: "float_field", kind: "scalar", T: 2 /*ScalarType.FLOAT*/ },
        // ...
```

Which isn't great. I'm sure the field metadata could have been designed with a bit better support for different naming conventions if that had been a design goal, but in general, with the reflection-based approach taken by protobuf-ts, supporting different naming conventions seems to come with increased complexity and/or code size. Happy to hear your thoughts on this.

**Comment by kskalski** [comment]
I understand that the current code allows to omit ``localName`` when ``name.toLowerCase() == localName`` (as in https://github.com/timostamm/protobuf-ts/blob/c559c0d920ee50b9b54d2f8fff6a7d59b46a3f50/packages/runtime/src/reflection-info.ts#L427)
And that ideally we would like to minimize the size of generated code and complexity of spec/reflection code base.

I can see following options here:
1. make the interpretation of fields in spec a bit more flexible, e.g. when ``localName`` is specified, but ``name`` isn't, then fill it in - ``field.name ??= field.localName``

2. with a bit different compatibility implications, similar to 1., but using a completely new field, e.g. ``commonName``

3. pass some kind of 'mode' constant to the reflection code that would say how the ``localName`` should be generated from ``name`` - it seems that's a larger, but doable addition of code in several places. If that is done, maybe also json name handling should be specified by its mode instead of passing the json names explicitly - this would actually reduce generated code size in other scenarios.

4. similar as 3., but maybe it would be fine to use a runtime global variable that reflection code would check - that should be a much smaller addition

5. simply accept that field info objects passed to reflection are a bit larger (when developer chooses to keep original names *and* given name actually differs from lower case one) and generate both ``name`` and ``localName`` as needed. Note that this variant is already used in case of JSON ``useProtoFieldName`` option.

Given that there is already a dedicated field for storing the interface field's name, it seems to me that this feature isn't particularly hard to add (especially the last implementation alternative). 

My intuition suggests the best option would be 3. / 4. where a kind of bit field flag would be used selecting one of the predefined approaches for naming ``localName`` and ``jsonName``, so generated code would define/pass e.g. ``NamingMode.LOCAL_USE_PROTO_NAME | NamingMode.JSON_USE_PROTO_NAME``.

**Comment by timostamm** [comment]
Global variables are not an option, it should definitely be possible to run two parts of code generated with different options in parallel. Disregarding breaking changes, I'm not sure how a naming-mode would play out with `useProtoFieldName` (which is a _runtime_ option, and part of the protocol buffers recommended implementation), and think it would need more investigation. However, I would prefer this not to be a breaking change, which leaves option 1 and 5. 

Option 1 will complicate the typings for field metadata, or make omission of `name` and `localName` a runtime error. A bit too much magic IMO.

Option 5 looks like the most reasonable approach to me, despite increased code size. Happy to take a PR for this.

**Comment by jcready** [comment]
I believe this has been addressed in https://github.com/timostamm/protobuf-ts/pull/250 which was released in [v2.3.0](https://github.com/timostamm/protobuf-ts/releases/tag/v2.3.0).
